cylc/flow/prerequisite.py

```diff
@@ -274,10 +280,12 @@ def satisfy_me(
             if output_tuple not in self._satisfied:
                 continue
             valid.add(output)
-            if self._satisfied[output_tuple] != 'satisfied naturally':
-                self[output_tuple] = (
-                    'force satisfied' if forced else 'satisfied naturally'
-                )
+            msg: SatisfiedState = (
+                'satisfied by skip mode' if mode == RunMode.SKIP
+                else 'satisfied naturally'
+            )
+            if self._satisfied[output_tuple] != msg:
+                self[output_tuple] = 'force satisfied' if forced else msg
```

Comment:
Should this be:

```python
msg: SatisfiedState
if forced:
    msg = 'force satisfied'
elif mode == RunMode.SKIP:
    msg = 'satisfied by skip mode'
else:
   msg = 'satisfied naturally'
if self._satisfied[output_tuple] != msg:
    self[output_tuple] = msg

Reply:
On 8.3.x, `forced` does not override an existing truthy satisfaction state: https://github.com/cylc/cylc-flow/blob/347921fdee5b3e8d0f553f81bb5220a8618caa84/cylc/flow/prerequisite.py#L261-L269

(N.B. this forcing is missing from the `satisfy_me()` method on 8.3.x; I added it in #6472 because it should be there for `cylc set --pre` which uses either `set_satisfied()` for `--pre all` or `satisfy_me()` for `--pre something-specific`)

**But,** if you [blame the lines above](https://github.com/cylc/cylc-flow/blame/347921fdee5b3e8d0f553f81bb5220a8618caa84/cylc/flow/prerequisite.py#L268), this "not override an existing truthy satisfaction state" behaviour is 7 years old and obviously at the time had nothing to do with `cylc set --pre`.

So it could be argued the correct behaviour is actually `forced` should override existing truthy satisfaction states, which would prevent `cylc remove` from unsetting the prereq.

But on the other hand it could be argued that `cylc set --pre already-satisfied` should do nothing.

🤷‍♂️

Reply:
Ok, that makes sense.

Unfortunately, it means that we are handling "force satisfied" and "satisfied by skip mode" differently.

* "force satisfied" **will not** override the previous state.
* "satisfied by skip mode" **will** override the previous state.

There are two possible consistent options:

1. The value a prerequisite is **first** satisfied is preserved.
2. The value a prerequisite is **last** satisfied overrides previous values.

Given the "force satisfied" value follows option (1), something which makes reasonable sense from a provenance perspective, we should make the "satisfied by skip mode" value behave the same.

Reply:
```suggestion
            if not self._satisfied[output_tuple]:
                self[output_tuple] = (
                    'force satisfied' if forced
                    else 'satisfied by skip mode' if mode == RunMode.SKIP
                    else 'satisfied naturally'
                )
```

Reply:
Realised the tests could do with updating too, so here is another auxiliary PR! https://github.com/wxtim/cylc/pull/71

Reply:
Crikey Ronnie - Another case of review by having half written something then Ronnie then puts up as a PR to a PR, but good work if we can get this out before Christmas.